From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sat, 17 May 2025 17:36:22 +0800
Subject: [PATCH] Replay Mod API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 986a34b5b7872ba964f2f9cdd735439d91973557..09eda069b9d82ce953d8370d7df5afe581341173 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3017,4 +3017,8 @@ public final class Bukkit {
         return server.getBotManager();
     }
     // Leaves end - Bot API
+
+    public static @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager() { // Leaves - Photographer API
+        return server.getPhotographerManager();
+    }
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1a81b53408cd20af750d507fab8d037de389f450..471297bd2c74430a2b4a5b7bb93521326e1a2a9d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2789,4 +2789,6 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.leavesmc.leaves.entity.bot.BotManager getBotManager();
     // Leaves end - Bot API
+
+    @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager(); // Leaves - Photographer API
 }
