From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:59:11 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 3b489ea7933b7275a1b3e65820154cedc8a61768..fb615407be4b6726ec9e412209e92cd9809474a0 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3029,4 +3029,10 @@ public final class Bukkit {
         return server.getBytebufAllocator();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    public static org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return server.getLeavesConfig();
+    }
+    // Leaves end - Config API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d2b60183d8a84d8cf3faeec0dfa8528a747d32dc..00a6c7fd0daa98ea12cb886274d70af1e5b4f1a9 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2797,4 +2797,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Bytebuf API
     org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator();
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig();
+    // Leaves End - Config API
 }
