From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 26 Jan 2025 13:58:49 +0800
Subject: [PATCH] Build changes


diff --git a/ca/spottedleaf/moonrise/paper/PaperHooks.java b/ca/spottedleaf/moonrise/paper/PaperHooks.java
index 4a3f07daad8b48eaf3bea02bdacbfb0670dddf90..e1e032e709193b5a8008d69b748bc54bc660eb64 100644
--- a/ca/spottedleaf/moonrise/paper/PaperHooks.java
+++ b/ca/spottedleaf/moonrise/paper/PaperHooks.java
@@ -29,7 +29,7 @@ import net.minecraft.world.phys.AABB;
 import java.util.List;
 import java.util.function.Predicate;
 
-public final class PaperHooks extends BaseChunkSystemHooks implements PlatformHooks {
+public class PaperHooks extends BaseChunkSystemHooks implements PlatformHooks { // Leaves - not final
 
     private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger();
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c5dc23ff9a3cd30f7bae8ca896fede1a19e7a901..3fd785f2094ea8da37fe454ecc773ccffa4fa7c3 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1318,7 +1318,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 LOGGER.info("*************************************************************************************");
                 LOGGER.info("This is the first time you're starting this server.");
                 LOGGER.info("It's recommended you read our 'Getting Started' documentation for guidance.");
-                LOGGER.info("View this and more helpful information here: https://docs.papermc.io/paper/next-steps");
+                LOGGER.info("View this and more helpful information here: https://docs.leavesmc.org/leaves/guides/next-steps"); // Leaves - change
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
@@ -1974,7 +1974,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return org.leavesmc.leaves.LeavesConfig.mics.serverModName; // Leaves - Custom // Paper
     }
 
     public ServerClockManager clockManager() {
diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index b810cb0a9dd8713aed70cc9fdeea1bc249c8e0dc..82cc8fe72621113c514c569d56a9c9a0958df5b6 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -183,7 +183,7 @@ public class MinecraftServerGui extends JComponent {
 
     // Paper start - Add onboarding message for initial server start
     private JComponent buildOnboardingPanel() {
-        String onboardingLink = "https://docs.papermc.io/paper/next-steps";
+        String onboardingLink = "https://docs.leavesmc.org/leaves/guides/next-steps"; // Leaves - change
         JPanel jPanel = new JPanel();
 
         javax.swing.JLabel jLabel = new javax.swing.JLabel("If you need help setting up your server you can visit:");
