From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 3 Apr 2022 11:31:04 +0800
Subject: [PATCH] Movable Budding Amethyst


diff --git a/net/minecraft/world/level/block/BuddingAmethystBlock.java b/net/minecraft/world/level/block/BuddingAmethystBlock.java
index 15d58eb7b6a24d28e1ed092f654db62c5230c61f..e2ceea06d973706ad9ddb962c5acd439a3531ed7 100644
--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -51,4 +51,12 @@ public class BuddingAmethystBlock extends AmethystBlock {
     public static boolean canClusterGrowAtState(final BlockState state) {
         return state.isAir() || state.is(Blocks.WATER) && state.getFluidState().isFull();
     }
+
+    // Leaves start - budding amethyst can push by piston
+    @Override
+    @org.jetbrains.annotations.Nullable
+    public net.minecraft.world.level.material.PushReaction getResetPushReaction() {
+        return org.leavesmc.leaves.LeavesConfig.modify.movableBuddingAmethyst ? net.minecraft.world.level.material.PushReaction.PUSH_PULL : null;
+    }
+    // Leaves end - budding amethyst can push by piston
 }
