From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:05:23 +0800
Subject: [PATCH] Old hopper suckin behavior


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index ce6b64e00a0a13ebdceda56d975d79f60c26b86d..064ef5c59b52344bdc0ab6f4df46e17669f576c3 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -492,8 +492,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             return false;
         } else {
             boolean isBlocked = hopper.isGridAligned()
-                && blockState.isCollisionShapeFullBlock(level, blockPos)
-                && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS);
+                && (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldHopperSuckInBehavior && blockState.isCollisionShapeFullBlock(level, blockPos))
+                && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS); // Leaves - oldHopperSuckInBehavior
             if (!isBlocked) {
                 for (ItemEntity entity : getItemsAtAndAbove(level, hopper)) {
                     if (addItem(hopper, entity)) {
