From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 30 Jul 2023 12:20:16 +0800
Subject: [PATCH] Zero tick plants


diff --git a/net/minecraft/world/level/block/BambooStalkBlock.java b/net/minecraft/world/level/block/BambooStalkBlock.java
index 57f0ab50c2eb22701bc709baba5a1e053757bfa1..da58399b39e9e3fdef179eb2acbd2785da365e8c 100644
--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -101,6 +101,10 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index 687f7bca0a546bb4dc7942194a76088a07e36770..9b4e9fe1cdffeed339f57f15cb95691dce30a3c9 100644
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -40,6 +40,10 @@ public class CactusBlock extends Block {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/ChorusFlowerBlock.java b/net/minecraft/world/level/block/ChorusFlowerBlock.java
index 4b2f1868ebd07b97a4900ea55aa31226a2d2e263..43d227f20018df818cde01ad0b90810dac57b2dd 100644
--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -36,6 +36,10 @@ public class ChorusFlowerBlock extends Block {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
index 4fd9e6bb598afdebfda4828e9f18afe556057f63..bad9d3d32aaeefbc0a72b88edb92ae5091b73861 100644
--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -156,4 +156,15 @@ public abstract class GrowingPlantHeadBlock extends GrowingPlantBlock implements
     protected GrowingPlantHeadBlock getHeadBlock() {
         return this;
     }
+
+    // Leaves start - zero tick plants
+    @Override
+    protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (!state.canSurvive(level, pos)) {
+            level.destroyBlock(pos, true);
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        }
+    }
+    // Leaves end - zero tick plants
 }
diff --git a/net/minecraft/world/level/block/SugarCaneBlock.java b/net/minecraft/world/level/block/SugarCaneBlock.java
index b7bc0151fa6b01ed4814ca2050daa63f5b5b05a1..393c0b1631efc0f686a520fefa816a45773103e5 100644
--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -36,6 +36,10 @@ public class SugarCaneBlock extends Block {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
