From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 14 Sep 2023 20:23:03 +0800
Subject: [PATCH] Force Void Trade


diff --git a/net/minecraft/world/entity/npc/villager/AbstractVillager.java b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
index 5d742dcd507a1f47feb802ad5c2ce24a05d57bbf..7de22767667847d24f23de9eec4eb1a28434ba0c 100644
--- a/net/minecraft/world/entity/npc/villager/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
@@ -58,6 +58,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     private @Nullable Player tradingPlayer;
     protected @Nullable MerchantOffers offers;
     private final SimpleContainer inventory = new SimpleContainer(8, (org.bukkit.craftbukkit.entity.CraftAbstractVillager) this.getBukkitEntity()); // CraftBukkit - add argument
+    protected boolean voidTrade = false; // Leaves - force void trade
 
     public AbstractVillager(final EntityType<? extends AbstractVillager> type, final Level level) {
         super(type, level);
@@ -149,7 +150,13 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     @Override
     public void processTrade(MerchantOffer offer, io.papermc.paper.event.player.@Nullable PlayerPurchaseEvent event) { // The MerchantRecipe passed in here is the one set by the PlayerPurchaseEvent
         if (event == null || event.willIncreaseTradeUses()) {
-            offer.increaseUses();
+            // Leaves start - force void trade
+            if (this.voidTrade) {
+                offer.increaseVoidUses();
+            } else {
+                offer.increaseUses();
+            }
+            // Leaves end - force void trade
         }
         if (event == null || event.isRewardingExp()) {
             this.rewardTradeXp(offer);
@@ -161,7 +168,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     @Override
     public void notifyTrade(final MerchantOffer offer) {
         // offer.increaseUses(); // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
-        this.ambientSoundTime = -this.getAmbientSoundInterval();
+        if (!this.voidTrade) this.ambientSoundTime = -this.getAmbientSoundInterval(); // Leaves - force void trade
         // this.rewardTradeXp(offer); // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
         if (this.tradingPlayer instanceof ServerPlayer) {
             CriteriaTriggers.TRADE.trigger((ServerPlayer)this.tradingPlayer, this, offer.getResult());
@@ -178,7 +185,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     @Override
     public void notifyTradeUpdated(final ItemStack itemStack) {
         if (!this.level().isClientSide() && this.ambientSoundTime > -this.getAmbientSoundInterval() + 20) {
-            this.ambientSoundTime = -this.getAmbientSoundInterval();
+            if (!this.voidTrade) this.ambientSoundTime = -this.getAmbientSoundInterval(); // Leaves - force void trade
             this.makeSound(this.getTradeUpdatedSound(!itemStack.isEmpty()));
         }
     }
@@ -220,9 +227,21 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     }
 
     protected void stopTrading() {
+        // Leaves start - force void trade
+        this.voidTrade = false;
+        if (this.offers != null) {
+            this.offers.forEach(MerchantOffer::resetVoidUses);
+        }
+        // Leaves end - force void trade
         this.setTradingPlayer(null);
     }
 
+    // Leaves start - force void trade
+    public void setVoidTrade() {
+        this.voidTrade = true;
+    }
+    // Leaves end - force void trade
+
     @Override
     public void die(final DamageSource source) {
         super.die(source);
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 5096dee853cd0b74e62cd90b2638a3235c378e00..26d4c5ebe5b070b27351c48d98fe7782bff08a59 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -304,6 +304,7 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
     }
 
     private void startTrading(final Player player) {
+        this.voidTrade = false; // Leaves - force void trade
         this.updateSpecialPrices(player);
         this.setTradingPlayer(player);
         this.openTradingScreen(player, this.getDisplayName(), this.getVillagerData().level());
@@ -568,8 +569,12 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
     @Override
     protected void rewardTradeXp(final MerchantOffer offer) {
         int popXp = 3 + this.random.nextInt(4);
-        this.villagerXp = this.villagerXp + offer.getXp();
-        this.lastTradedPlayer = this.getTradingPlayer();
+        // Leaves start - force void trade
+        if (!this.voidTrade) {
+            this.villagerXp += offer.getXp();
+            this.lastTradedPlayer = this.getTradingPlayer();
+        }
+        // Leaves end - force void trade
         if (this.shouldIncreaseLevel()) {
             if (this.level() instanceof ServerLevel serverLevel) {
                 this.increaseMerchantCareer(serverLevel);
diff --git a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
index 7af7514a204914957de02d8a775fb8ad9ea6d010..647bbcacd36298b27c6ef94b6c9e6281e29fe927 100644
--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
@@ -120,6 +120,7 @@ public class WanderingTrader extends AbstractVillager implements Consumable.Over
                     return InteractionResult.CONSUME;
                 }
 
+                this.voidTrade = false; // Leaves - force void trade
                 this.setTradingPlayer(player);
                 this.openTradingScreen(player, this.getDisplayName(), 1);
             }
diff --git a/net/minecraft/world/inventory/MerchantMenu.java b/net/minecraft/world/inventory/MerchantMenu.java
index 98cbe25749983c9187419a8df584727b64599bf0..2a2a514c8c6f0ab55246e8000791cb18aa1d9db2 100644
--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -26,7 +26,7 @@ public class MerchantMenu extends AbstractContainerMenu {
     private static final int SELLSLOT2_X = 162;
     private static final int BUYSLOT_X = 220;
     private static final int ROW_Y = 37;
-    private final Merchant trader;
+    public final Merchant trader; // Leaves - private -> public
     private final MerchantContainer tradeContainer;
     private int merchantLevel;
     private boolean showProgressBar;
diff --git a/net/minecraft/world/item/trading/MerchantOffer.java b/net/minecraft/world/item/trading/MerchantOffer.java
index b7766e7155a6a5345bc4eefcbd05e6848e1bf026..3c009f5debb02f89e4de9cee44b7dd8e1971feb9 100644
--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -39,6 +39,7 @@ public class MerchantOffer {
     public float priceMultiplier;
     public int xp;
     public boolean ignoreDiscounts; // Paper - Add ignore discounts API
+    public int voidTradeUses; // Leaves - force void trade
 
     // CraftBukkit start
     private org.bukkit.craftbukkit.inventory.@org.jspecify.annotations.Nullable CraftMerchantRecipe bukkitHandle;
@@ -78,6 +79,7 @@ public class MerchantOffer {
         this.priceMultiplier = priceMultiplier;
         this.xp = xp;
         this.ignoreDiscounts = ignoreDiscounts; // Paper
+        this.voidTradeUses = 0; // Leaves - force void trade
     }
 
     public MerchantOffer(final ItemCost buy, final ItemStack result, final int maxUses, final int xp, final float priceMultiplier) {
@@ -186,6 +188,16 @@ public class MerchantOffer {
         this.uses++;
     }
 
+    // Leaves start - force void trade
+    public void increaseVoidUses() {
+        ++this.voidTradeUses;
+    }
+
+    public void resetVoidUses() {
+        this.voidTradeUses = 0;
+    }
+    // Leaves end - force void trade
+
     public int getDemand() {
         return this.demand;
     }
@@ -215,7 +227,7 @@ public class MerchantOffer {
     }
 
     public boolean isOutOfStock() {
-        return this.uses >= this.maxUses;
+        return this.uses + this.voidTradeUses >= this.maxUses; // Leaves - force void trade
     }
 
     public void setToOutOfStock() {
diff --git a/net/minecraft/world/level/block/EndGatewayBlock.java b/net/minecraft/world/level/block/EndGatewayBlock.java
index 4ac1da84017191eb0d415ff0a4ed934538da985d..022584b52328ef4bfbd2a53282c7271c0f4b3b9f 100644
--- a/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -104,6 +104,15 @@ public class EndGatewayBlock extends BaseEntityBlock implements Portal {
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
         if (currentLevel.getBlockEntity(portalEntryPos) instanceof TheEndGatewayBlockEntity endGatewayBlockEntity) {
             Vec3 teleportPosition = endGatewayBlockEntity.getPortalPosition(currentLevel, portalEntryPos);
+            // Leaves start - force void trade
+            if (org.leavesmc.leaves.LeavesConfig.modify.forceVoidTrade && teleportPosition != null && entity instanceof net.minecraft.server.level.ServerPlayer player) {
+                if (player.containerMenu instanceof net.minecraft.world.inventory.MerchantMenu merchantMenu) {
+                    if (merchantMenu.trader instanceof net.minecraft.world.entity.npc.villager.AbstractVillager villager) {
+                        villager.setVoidTrade();
+                    }
+                }
+            }
+            // Leaves end - force void trade
             if (teleportPosition == null) {
                 return null;
             } else {
