From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Thu, 12 Jun 2025 12:55:12 +0800
Subject: [PATCH] Leaves I18n support


diff --git a/net/minecraft/locale/Language.java b/net/minecraft/locale/Language.java
index 3ee139b4c47cd5a747cf7aa4f95beb0c67f86143..22289e4247082c6c2388ab5df686f89e708a8eb1 100644
--- a/net/minecraft/locale/Language.java
+++ b/net/minecraft/locale/Language.java
@@ -37,6 +37,7 @@ public abstract class Language {
         Map<String, String> loadedData = new HashMap<>();
         BiConsumer<String, String> output = loadedData::put;
         parseTranslations(output, "/assets/minecraft/lang/en_us.json");
+        org.leavesmc.leaves.util.ServerI18nUtil.loadLeavesI18nDefault(output); // Leaves I18n
         deprecatedInfo.applyToMap(loadedData);
         final Map<String, String> storage = Map.copyOf(loadedData);
         return new Language() {
@@ -66,7 +67,7 @@ public abstract class Language {
         };
     }
 
-    private static void parseTranslations(final BiConsumer<String, String> output, final String path) {
+    public static void parseTranslations(final BiConsumer<String, String> output, final String path) { // Leaves - private -> public
         try (InputStream stream = Language.class.getResourceAsStream(path)) {
             loadFromJson(stream, output);
         } catch (IOException | JsonParseException e) {
diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index b730c6694e819813f8eb08b742d660937f9cbc21..f68b38e274947f96b371d832377e4b7c06e0c13e 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -154,6 +154,8 @@ public class Main {
                 return;
             }
 
+            org.leavesmc.leaves.util.ServerI18nUtil.preInit(); // Leaves I18n
+
             // Paper start - Detect headless JRE
             String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
             if (awtException != null) {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index c251bb40f9e2b370d38f1b4f0b8f7b18f88d2470..647abeb9e487cdead07cb38ae0c0184047923e6e 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -292,6 +292,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
         org.leavesmc.leaves.LeavesConfig.init((java.io.File) options.valueOf("leaves-settings")); // Leaves - Server Config
         this.getBotList().loadResumeBotInfo(); // Leaves - load resident bot info
+        org.leavesmc.leaves.util.ServerI18nUtil.init(); // Leaves I18n
 
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
