From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Mon, 29 Sep 2025 10:24:45 +0800
Subject: [PATCH] Do not check NaN values in entities


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 4b9b05d6d5c74bbd92169f14f06e936690c90f4a..6b2e71febce509d3eb8aa4f53a8454d3007adb19 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -2581,9 +2581,11 @@ public abstract class Entity
             }
             } // CraftBukkit
 
+            // Leaves start - Do not check NaN values
             output.store("Motion", Vec3.CODEC, this.getDeltaMovement());
             // CraftBukkit start - Checking for NaN pitch/yaw and resetting to zero
             // TODO: make sure this is the best way to address this.
+            /*
             if (Float.isNaN(this.yRot)) {
                 this.yRot = 0;
             }
@@ -2591,7 +2593,9 @@ public abstract class Entity
             if (Float.isNaN(this.xRot)) {
                 this.xRot = 0;
             }
+            */
             // CraftBukkit end
+            // Leaves end - Do not check NaN values
             output.store("Rotation", Vec2.CODEC, new Vec2(this.getYRot(), this.getXRot()));
             output.putDouble("fall_distance", this.fallDistance);
             output.putShort("Fire", (short)this.remainingFireTicks);
