From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Thu, 15 May 2025 12:17:30 +0800
Subject: [PATCH] Vanilla portal handle


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 4b9b05d6d5c74bbd92169f14f06e936690c90f4a..0b304eee945c64b75d0e6cd6a5ad76c6d6bdf454 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -863,6 +863,7 @@ public abstract class Entity
 
     // CraftBukkit start
     public void postTick() {
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle) return; // Leaves - vanilla
         // No clean way to break out of ticking once the entity has been copied to a new world, so instead we move the portalling later in the tick cycle
         if (!(this instanceof ServerPlayer) && this.isAlive()) { // Paper - don't attempt to teleport dead entities
             this.handlePortal();
@@ -884,7 +885,7 @@ public abstract class Entity
             this.boardingCooldown--;
         }
 
-        if (this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle || this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick // Leaves - vanilla
         if (this.canSpawnSprintParticle()) {
             this.spawnSprintParticle();
         }
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index 1c043c1136fc42e888125047ed73b1dc32276183..28ee5a62cfd584707c879e74ff87792db208eca9 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -307,7 +307,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
         this.checkBelowWorld();
         this.computeSpeed();
-        // this.handlePortal(); // CraftBukkit - handled in postTick
+        if (org.leavesmc.leaves.LeavesConfig.fix.vanillaPortalHandle) this.handlePortal(); // CraftBukkit - handled in postTick // Leaves - vanilla
         this.behavior.tick();
         // CraftBukkit start
         org.bukkit.World bworld = this.level().getWorld();
