From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fortern <blueten.ki@gmail.com>
Date: Sat, 20 Jan 2024 02:50:56 +0800
Subject: [PATCH] Placing locked hopper no longer send NC updates


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 0c9cbb2e65c711ab8e46c77aeab7a580a977a72d..1075ab28e7cf4434958ca5183213e2172c71e76b 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1181,7 +1181,11 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
 
             if ((updateFlags & Block.UPDATE_NEIGHBORS) != 0) {
-                this.updateNeighborsAt(pos, oldState.getBlock());
+                // Leaves start - Placing locked hopper doesn't send NC updates.
+                if (!(blockState.getBlock() instanceof net.minecraft.world.level.block.HopperBlock) || blockState.getValue(net.minecraft.world.level.block.HopperBlock.ENABLED)) {
+                    this.updateNeighborsAt(pos, oldState.getBlock());
+                }
+                // Leaves end - Placing locked hopper doesn't send NC updates.
                 if (!this.isClientSide() && blockState.hasAnalogOutputSignal()) {
                     this.updateNeighbourForOutputSignal(pos, blockState.getBlock());
                 }
