From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:05:23 +0800
Subject: [PATCH] Old hopper suckin behavior


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 94d4bd6b37dd950d7360b04fd37b14fdac857bca..3512ea90a70b978f7e6558e66ce720a8b2d64727 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -568,8 +568,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             return false;
         } else {
             boolean isBlocked = hopper.isGridAligned()
-                && blockState.isCollisionShapeFullBlock(level, blockPos)
-                && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS);
+                && (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldHopperSuckInBehavior && blockState.isCollisionShapeFullBlock(level, blockPos))
+                && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS); // Leaves - oldHopperSuckInBehavior
             if (!isBlocked) {
                 for (ItemEntity entity : getItemsAtAndAbove(level, hopper)) {
                     if (addItem(hopper, entity)) {
