From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 6 Apr 2025 00:19:06 +0800
Subject: [PATCH] Disable packet limit


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 0f06d509125e72ec11d263d0d5c2179d65f6d8dd..f86eb7eac87830452b36e60e8c23d4da5f0030bf 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -221,8 +221,8 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             if (this.stopReadingPackets) {
                 return;
             }
-            if (this.allPacketCounts != null ||
-                io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass())) {
+                if (!org.leavesmc.leaves.LeavesConfig.modify.disablePacketLimit && (this.allPacketCounts != null || // Leaves - can disable
+                    io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass()))) { // Leaves - can disable
                 long time = System.nanoTime();
                 synchronized (PACKET_LIMIT_LOCK) {
                     if (this.allPacketCounts != null) {
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 3a9c568fe3ce1fd45678294c05785f5e18e19906..1af46d2fa1564ed5f516c8c32d50521758a68473 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -844,7 +844,7 @@ public class ServerGamePacketListenerImpl
     public void handleCustomCommandSuggestions(final ServerboundCommandSuggestionPacket packet) {
         // PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level()); // Paper - AsyncTabCompleteEvent; run this async
         // CraftBukkit start
-        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.nameAndId()) && !this.server.isSingleplayerOwner(this.player.nameAndId())) { // Paper - configurable tab spam limits
+        if (!org.leavesmc.leaves.LeavesConfig.modify.disablePacketLimit && !this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.nameAndId()) && !this.server.isSingleplayerOwner(this.player.nameAndId())) { // Paper - configurable tab spam limits // Leaves - can disable
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - Kick event cause // Paper - add proper async disconnect
             return;
         }
@@ -2098,6 +2098,7 @@ public class ServerGamePacketListenerImpl
     private long lastLimitedPacket = -1;
 
     private boolean checkLimit(long timestamp) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.disablePacketLimit) return true; // Leaves - disable
         if (!io.papermc.paper.configuration.GlobalConfiguration.get().spamLimiter.incomingPacketThreshold.enabled()) {
             return true;
         }
@@ -2653,6 +2654,7 @@ public class ServerGamePacketListenerImpl
 
     // Spigot start - spam exclusions
     private void detectRateSpam(String message) {
+        if (org.leavesmc.leaves.LeavesConfig.modify.disablePacketLimit) return; // Leaves - disable
         // CraftBukkit start - replaced with thread safe throttle
         if (org.spigotmc.SpigotConfig.enableSpamExclusions) {
             for (String exclude : org.spigotmc.SpigotConfig.spamExclusions) {
@@ -3401,7 +3403,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handlePlaceRecipe(final ServerboundPlaceRecipePacket packet) {
         // Paper start - auto recipe limit
-        if (!org.bukkit.Bukkit.isPrimaryThread()) {
+        if (!org.leavesmc.leaves.LeavesConfig.modify.disablePacketLimit && !org.bukkit.Bukkit.isPrimaryThread()) { // Leaves - can disable
             if (!this.recipeSpamPackets.isIncrementAndUnderThreshold()) {
                 this.disconnectAsync(net.minecraft.network.chat.Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause // Paper - add proper async disconnect
                 return;
