From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bluemangoo <chenfy2006@qq.com>
Date: Fri, 19 Jul 2024 15:04:03 +0800
Subject: [PATCH] Disable crystal-portal proximity check


diff --git a/net/minecraft/world/item/EndCrystalItem.java b/net/minecraft/world/item/EndCrystalItem.java
index eaf7b3f7e9b8135e0ebc1aa7f764f58b45dd3aa5..b0cf5265f1bad7e81b9de22a0c48e272f8c41d70 100644
--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -54,7 +54,7 @@ public class EndCrystalItem extends Item {
             level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, above);
             EnderDragonFight fight = ((ServerLevel)level).getDragonFight();
             if (fight != null) {
-                fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup
+                            fight.tryRespawn(); // fight.tryRespawn(above); // Paper - Perf: Do crystal-portal proximity check before entity lookup // Leaves
             }
         }
 
diff --git a/net/minecraft/world/level/dimension/end/EnderDragonFight.java b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
index 86bfcae47d216f7041284f7a40d592d997b514b9..fa69eb989d826140f3cf7ad34701e1b0c24f818a 100644
--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -615,12 +615,14 @@ public class EnderDragonFight extends SavedData {
     }
 
     public boolean tryRespawn() { // CraftBukkit - return boolean
+        /* Leaves
         // Paper start - Perf: Do crystal-portal proximity check before entity lookup
         return this.tryRespawn(null);
     }
 
     public boolean tryRespawn(@Nullable final BlockPos placedEndCrystalPos) { // placedEndCrystalPos is null if the tryRespawn() call was not caused by a placed end crystal
         // Paper end - Perf: Do crystal-portal proximity check before entity lookup
+         */
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos location = this.exitPortalLocation;
             if (location == null) {
@@ -635,6 +637,7 @@ public class EnderDragonFight extends SavedData {
 
                 location = this.exitPortalLocation;
             }
+            /* Leaves
             // Paper start - Perf: Do crystal-portal proximity check before entity lookup
             if (placedEndCrystalPos != null && !this.level.paperConfig().misc.allowRemoteEnderDragonRespawning) {
                 int dy = placedEndCrystalPos.getY() - location.getY();
@@ -649,6 +652,7 @@ public class EnderDragonFight extends SavedData {
                 }
             }
             // Paper end - Perf: Do crystal-portal proximity check before entity lookup
+             */
 
             List<EndCrystal> crystals = Lists.newArrayList();
             BlockPos center = location.above(1);
