From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 3 Feb 2025 15:36:21 +0800
Subject: [PATCH] Return nether portal fix


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 8a53a7d733cca5e9be903255d7af05522dc6ce3a..5500c3efc711b340b62a7abb524e74017c3c0bc3 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1591,6 +1591,19 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         org.bukkit.event.player.PlayerChangedWorldEvent changeEvent = new org.bukkit.event.player.PlayerChangedWorldEvent(this.getBukkitEntity(), oldLevel.getWorld());
         this.level().getCraftServer().getPluginManager().callEvent(changeEvent);
         // CraftBukkit end
+                // Leaves start - nether portal fix
+                if (org.leavesmc.leaves.LeavesConfig.modify.netherPortalFix) {
+                    final net.minecraft.resources.ResourceKey<Level> fromDim = oldLevel.dimension();
+                    final net.minecraft.resources.ResourceKey<Level> toDim = this.level().dimension();
+                    if ((fromDim == Level.OVERWORLD && toDim == Level.NETHER) || (fromDim == Level.NETHER && toDim == Level.OVERWORLD)) {
+                        BlockPos fromPortal = org.leavesmc.leaves.util.ReturnPortalManager.findPortalAt(this, fromDim, this.lastPos);
+                        BlockPos toPos = this.blockPosition();
+                        if (fromPortal != null) {
+                            org.leavesmc.leaves.util.ReturnPortalManager.storeReturnPortal(this, toDim, toPos, fromPortal);
+                        }
+                    }
+                }
+                // Leaves end - nether portal fix
         // Paper start - Reset shield blocking on dimension change
         if (this.isBlocking()) {
             this.stopUsingItem();
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 0d9ace52f7ecd6af3aafd3dc629f33d5bad24116..fc0476948858a0b87d5df5ee591e27c55e5fd5d4 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -695,6 +695,18 @@ public abstract class PlayerList {
         // It's possible for respawn to be in a diff dimension
         if (fromLevel != level) {
             new org.bukkit.event.player.PlayerChangedWorldEvent(player.getBukkitEntity(), fromLevel.getWorld()).callEvent();
+            // Leaves start - nether portal fix
+            if (org.leavesmc.leaves.LeavesConfig.modify.netherPortalFix) {
+                final net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> fromDim = fromLevel.dimension();
+                final net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> toDim = level.dimension();
+                if ((fromDim == net.minecraft.world.level.Level.OVERWORLD && toDim == net.minecraft.world.level.Level.NETHER) || (fromDim == net.minecraft.world.level.Level.NETHER && toDim == net.minecraft.world.level.Level.OVERWORLD)) {
+                    net.minecraft.core.BlockPos fromPortal = org.leavesmc.leaves.util.ReturnPortalManager.findPortalAt(player, fromDim, player.lastPos);
+                    if (fromPortal != null) {
+                        org.leavesmc.leaves.util.ReturnPortalManager.storeReturnPortal(player, toDim, player.blockPosition(), fromPortal);
+                    }
+                }
+            }
+            // Leaves end - nether portal fix
             player.triggerDimensionChangeTriggers(level);
         }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 7bc07a621a07280b7a728c349934d345338e4ef0..a328dc8191e8c8e212eadc4f2ea269ec86cfcf85 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -261,7 +261,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public int useItemRemaining;
     protected int fallFlyTicks;
     private long lastKineticHitFeedbackTime = -2147483648L;
-    private BlockPos lastPos;
+    public BlockPos lastPos; // Leaves - private -> public
     private Optional<BlockPos> lastClimbablePos = Optional.empty();
     private @Nullable DamageSource lastDamageSource;
     private long lastDamageStamp;
diff --git a/net/minecraft/world/level/block/NetherPortalBlock.java b/net/minecraft/world/level/block/NetherPortalBlock.java
index 2838123f1e8fe538c36ff3707e39f7a4657ba0de..0f3518e864c3917a0b9b173beff37af16fc0dace 100644
--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -188,7 +188,18 @@ public class NetherPortalBlock extends Block implements Portal {
         final WorldBorder worldBorder
         , org.bukkit.craftbukkit.event.PortalEventResult result // CraftBukkit
     ) {
-        Optional<BlockPos> exitPortalPos = newLevel.getPortalForcer().findClosestPortalPosition(approximateExitPos, worldBorder, result.searchRadius()); // CraftBukkit
+        // Leaves start - fix return portal
+        Optional<BlockPos> exitPortalPos = Optional.empty();
+        if (org.leavesmc.leaves.LeavesConfig.modify.netherPortalFix && entity instanceof net.minecraft.server.level.ServerPlayer player) {
+            org.leavesmc.leaves.util.ReturnPortalManager.ReturnPortal portal = org.leavesmc.leaves.util.ReturnPortalManager.findReturnPortal(player, entity.level().dimension(), entity.blockPosition());
+            if (portal != null && newLevel.getBlockState(portal.pos()).is(Blocks.NETHER_PORTAL)) {
+                exitPortalPos = Optional.of(portal.pos());
+            }
+        }
+        if (exitPortalPos.isEmpty()) {
+            exitPortalPos = newLevel.getPortalForcer().findClosestPortalPosition(approximateExitPos, worldBorder, result.searchRadius()); // CraftBukkit
+        }
+        // Leaves end - fix return portal
         BlockUtil.FoundRectangle exitPortal;
         TeleportTransition.PostTeleportTransition post;
         if (exitPortalPos.isPresent()) {
