From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 19 May 2025 00:06:03 +0800
Subject: [PATCH] Spawn invulnerable time


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 394b29a5cc1cd04a3accbf3de82142c547290473..b9045349571c3d0fb399244bb757e4dc082d6dfe 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -245,6 +245,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     private int lastSentFood = -99999999;
     private boolean lastFoodSaturationZero = true;
     public int lastSentExp = -99999999;
+    protected int spawnInvulnerableTime = 60; // Leaves - spawn invulnerable time
     private ChatVisiblity chatVisibility = ChatVisiblity.FULL;
     public ParticleStatus particleStatus = ParticleStatus.ALL;
     private boolean canChatColor = true;
@@ -739,6 +740,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.resetOperationCountPerTick(); // Leaves - player operation limiter
         this.gameMode.tick();
         this.wardenSpawnTracker.tick();
+        if (this.spawnInvulnerableTime > 0) --this.spawnInvulnerableTime; // Leaves - spawn invulnerable time
         if (this.invulnerableTime > 0) {
             this.invulnerableTime--;
         }
@@ -1323,6 +1325,14 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             return false;
         }
 
+        // Leaves start - spawn invulnerable time
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.spawnInvulnerableTime) {
+            if (this.spawnInvulnerableTime > 0 && !source.is(net.minecraft.tags.DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+                return false;
+            }
+        }
+        // Leaves end - spawn invulnerable time
+
         Entity entity = source.getEntity();
         if (!( // Paper - split the if statement. If below statement is false, hurtServer would not have been evaluated. Return false.
             !(entity instanceof Player player && !this.canHarmPlayer(player))
