From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Tue, 22 Apr 2025 23:32:41 +0800
Subject: [PATCH] Leaves Utils


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a467fa895772c9459ebfb320a1f35485409ce9f9..15a01417a72e2eddcd517652a9680989f04191d0 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -388,6 +388,7 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    private net.minecraft.nbt.CompoundTag leavesData = new net.minecraft.nbt.CompoundTag(); // Leaves - Leaves ex data
 
     public void inactiveTick() {
     }
@@ -2716,6 +2717,7 @@ public abstract class Entity
                 output.putBoolean("Paper.FreezeLock", true);
             }
             // Paper end
+            output.store("Leaves.Data", net.minecraft.nbt.CompoundTag.CODEC, leavesData); // Leaves - leaves ex data
         } catch (Throwable t) {
             CrashReport report = CrashReport.forThrowable(t, "Saving entity NBT");
             CrashReportCategory category = report.addCategory("Entity being saved");
@@ -2839,6 +2841,7 @@ public abstract class Entity
             }
             freezeLocked = input.getBooleanOr("Paper.FreezeLock", false);
             // Paper end
+            leavesData = input.read("Leaves.Data", net.minecraft.nbt.CompoundTag.CODEC).orElse(new net.minecraft.nbt.CompoundTag());
         } catch (Throwable t) {
             CrashReport report = CrashReport.forThrowable(t, "Loading entity NBT");
             CrashReportCategory category = report.addCategory("Entity being loaded");
@@ -4623,7 +4626,7 @@ public abstract class Entity
         return this.is(EntityTypeTags.DEFLECTS_PROJECTILES) ? ProjectileDeflection.REVERSE : ProjectileDeflection.NONE;
     }
 
-    public @Nullable LivingEntity getControllingPassenger() {
+    public @org.jetbrains.annotations.NotNull LivingEntity getControllingPassenger() { // Leaves - notnull
         return null;
     }
 
@@ -5436,4 +5439,10 @@ public abstract class Entity
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    // Leaves start - leaves ex data
+    public net.minecraft.nbt.CompoundTag getLeavesData() {
+        return leavesData;
+    }
+    // Leaves end - leaves ex data
 }
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 1d78eca5e060668baf5c41ec4a76dfdd94b7ee83..961bc61685e2b6f6d38aeb770018beb634378427 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -928,7 +928,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     }
 
     @Override
-    public @Nullable MinecraftServer getServer() {
+    public @org.jetbrains.annotations.NotNull MinecraftServer getServer() { // Leaves - notnull
         return null;
     }
 
diff --git a/net/minecraft/world/level/LevelAccessor.java b/net/minecraft/world/level/LevelAccessor.java
index a82be0511b6381a2e9ee32f9f5ba1cca123b0835..ecb3073113bdef133932df14ae67d20f34e248ef 100644
--- a/net/minecraft/world/level/LevelAccessor.java
+++ b/net/minecraft/world/level/LevelAccessor.java
@@ -42,7 +42,7 @@ public interface LevelAccessor extends CommonLevelAccessor, ScheduledTickAccess
         return this.getLevelData().getGameTime();
     }
 
-    @Nullable MinecraftServer getServer();
+    @org.jetbrains.annotations.NotNull MinecraftServer getServer(); // Leaves - notnull
 
     default Difficulty getDifficulty() {
         return this.getLevelData().getDifficulty();
diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index a8625939a4220050990b09097d5e3bfd2d8473d5..70960c3c280eedc3624f20139a52d5d4961d6217 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -699,6 +699,13 @@ public class Block extends BlockBehaviour implements ItemLike {
     }
     // CraftBukkit end
 
+    // Leaves start - reset push reaction
+    @org.jetbrains.annotations.Nullable
+    public net.minecraft.world.level.material.PushReaction getResetPushReaction() {
+        return null;
+    }
+    // Leaves end - reset push reaction
+
     private record ShapePairKey(VoxelShape first, VoxelShape second) {
         @Override
         public boolean equals(final Object o) {
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 7093613a09a79d7ffce34f1cb54ec2605c5af384..6be4d5c7801937e865788e08789de2814c6fa19d 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -785,7 +785,7 @@ public abstract class BlockBehaviour implements FeatureElement {
         }
 
         public PushReaction getPistonPushReaction() {
-            return !this.isDestroyable() ? PushReaction.BLOCK : this.pushReaction; // Paper - Protect Bedrock and End Portal/Frames from being destroyed
+            return !this.isDestroyable() ? PushReaction.BLOCK : this.getBlock().getResetPushReaction() == null ? this.pushReaction : this.getBlock().getResetPushReaction(); // Paper - Protect Bedrock // Leaves - reset push reaction
         }
 
         public boolean isSolidRender() {
