From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+lumine1909@users.noreply.github.com>
Date: Wed, 26 Jun 2024 17:59:56 +0800
Subject: [PATCH] Allow grindstone overstacking


diff --git a/net/minecraft/world/SimpleContainer.java b/net/minecraft/world/SimpleContainer.java
index 8cff3f5fd8e2c1bde8fc6066c42b29535e0ad258..f83877e92f5d0c67ad90d61d39f8cb80e8db0cca 100644
--- a/net/minecraft/world/SimpleContainer.java
+++ b/net/minecraft/world/SimpleContainer.java
@@ -193,6 +193,12 @@ public class SimpleContainer implements Container, StackedContentsCompatible {
     @Override
     public void setItem(final int slot, final ItemStack itemStack) {
         this.items.set(slot, itemStack);
+        // Leaves start - grindstone overstacking
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && org.leavesmc.leaves.util.ItemOverstackUtils.CurseEnchantedBook.isCursedEnchantedBook(itemStack)) {
+            this.setChanged();
+            return;
+        }
+        // Leaves end - grindstone overstacking
         itemStack.limitSize(this.getMaxStackLeaves(itemStack)); // Leaves - item over-stack util
         this.setChanged();
     }
diff --git a/net/minecraft/world/entity/vehicle/ContainerEntity.java b/net/minecraft/world/entity/vehicle/ContainerEntity.java
index 15692744697043ab68a7247534614afac8f781f8..0ffb5fb16d0a0bc529bcca9d8beae569edfc6d71 100644
--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -160,6 +160,11 @@ public interface ContainerEntity extends Container, MenuProvider {
     default void setChestVehicleItem(final int slot, final ItemStack itemStack) {
         this.unpackChestVehicleLootTable(null);
         this.getItemStacks().set(slot, itemStack);
+        // Leaves start - grindstone overstacking
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && org.leavesmc.leaves.util.ItemOverstackUtils.CurseEnchantedBook.isCursedEnchantedBook(itemStack)) {
+            return;
+        }
+        // Leaves end - grindstone overstacking
         itemStack.limitSize(this.getMaxStackLeaves(itemStack)); // Leaves - item over-stack util
     }
 
diff --git a/net/minecraft/world/inventory/AbstractContainerMenu.java b/net/minecraft/world/inventory/AbstractContainerMenu.java
index 4d953d4b3f6ba9b6b5c4fde6e07ba2603487bf4e..03dfef15b6fc81d4f4653cbeda41819189776a8c 100644
--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -601,7 +601,7 @@ public abstract class AbstractContainerMenu {
                 } else if (targetItemStack.isEmpty()) {
                     if (target.mayPlace(source)) {
                         int maxStackSize = target.getMaxStackSize(source);
-                        if (source.getCount() > maxStackSize) {
+                        if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && source.getCount() > maxStackSize) { // Leaves - grindstone overstacking
                             target.setByPlayer(source.split(maxStackSize));
                         } else {
                             inventory.setItem(buttonNum, ItemStack.EMPTY);
@@ -610,7 +610,7 @@ public abstract class AbstractContainerMenu {
                     }
                 } else if (target.mayPickup(player) && target.mayPlace(source)) {
                     int maxStackSize = target.getMaxStackSize(source);
-                    if (source.getCount() > maxStackSize) {
+                    if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && source.getCount() > maxStackSize) { // Leaves - grindstone overstacking
                         target.setByPlayer(source.split(maxStackSize));
                         target.onTake(player, targetItemStack);
                         if (!inventory.add(targetItemStack)) {
@@ -763,6 +763,10 @@ public abstract class AbstractContainerMenu {
         return this.moveItemStackTo(itemStack, startSlot, endSlot, backwards, false);
     }
     protected boolean moveItemStackTo(ItemStack itemStack, final int startSlot, final int endSlot, final boolean backwards, final boolean isCheck) { // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent - non-final itemStack
+        // Leaves - Add force move
+        return this.moveItemStackTo(itemStack, startSlot, endSlot, backwards, isCheck, false);
+    }
+    protected boolean moveItemStackTo(ItemStack itemStack, final int startSlot, final int endSlot, final boolean backwards, final boolean isCheck, final boolean forceMove) { // Leaves - Add force move
         if (isCheck) {
             itemStack = itemStack.copy();
         }
@@ -827,6 +831,14 @@ public abstract class AbstractContainerMenu {
                 // Paper end - Add PlayerTradeEvent and PlayerPurchaseEvent
                 if (target.isEmpty() && slot.mayPlace(itemStack)) {
                     int maxStackSize = slot.getMaxStackSize(itemStack);
+                    // Leaves start - Add force move
+                    if (forceMove) {
+                        slot.setByPlayer(itemStack.split(itemStack.getCount()));
+                        slot.setChanged();
+                        anythingChanged = true;
+                        break;
+                    }
+                    // Leaves end - Add force move
                     // Paper start - Add PlayerTradeEvent and PlayerPurchaseEvent
                     if (isCheck) {
                         itemStack.shrink(Math.min(itemStack.getCount(), maxStackSize));
diff --git a/net/minecraft/world/inventory/GrindstoneMenu.java b/net/minecraft/world/inventory/GrindstoneMenu.java
index e074c3f28599f650fab9bda25b69c596d6129781..04362beb334985e1ef9d407adb3b0374c6a019a7 100644
--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -182,7 +182,7 @@ public class GrindstoneMenu extends AbstractContainerMenu {
         int remaining = remaining1 + remaining2 + durability * 5 / 100;
         int count = 1;
         if (!input.isDamageableItem()) {
-            if (input.getMaxStackSize() < 2 || !ItemStack.matches(input, additional)) {
+            if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && input.getMaxStackSize() < 2 || !ItemStack.matches(input, additional)) { // Leaves - allowGrindstoneOverstacking
                 return ItemStack.EMPTY;
             }
 
@@ -252,7 +252,7 @@ public class GrindstoneMenu extends AbstractContainerMenu {
             ItemStack input = this.repairSlots.getItem(0);
             ItemStack additional = this.repairSlots.getItem(1);
             if (slotIndex == 2) {
-                if (!this.moveItemStackTo(item, 3, 39, true)) {
+                if (!this.moveItemStackTo(item, 3, 39, true, false, org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking)) { // Leaves - allowGrindstoneOverstacking: Disable stack check
                     return ItemStack.EMPTY;
                 }
 
diff --git a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 97b69ca8025321d4ffbdca80e593348b014a73d6..f6d97d4cbf071e4ddccdd3fef537e6a7a741b26a 100644
--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -392,7 +392,11 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
         ItemStack oldStack = this.items.get(slot);
         boolean same = !itemStack.isEmpty() && ItemStack.isSameItemSameComponents(oldStack, itemStack);
         this.items.set(slot, itemStack);
-        itemStack.limitSize(this.getMaxStackLeaves(itemStack)); // Leaves - item over-stack util
+        // Leaves start - grindstone overstacking
+        if (!org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking || !org.leavesmc.leaves.util.ItemOverstackUtils.CurseEnchantedBook.isCursedEnchantedBook(oldStack)) {
+            itemStack.limitSize(this.getMaxStackLeaves(itemStack)); // Leaves - item over-stack util
+        }
+        // Leaves end - grindstone overstacking
         if (slot == 0 && !same && this.level instanceof ServerLevel serverLevel) {
             this.cookingTotalTime = getTotalCookTime(serverLevel, this, this.recipeType, this.cookSpeedMultiplier); // Paper - cook speed multiplier API
             this.cookingTimer = 0;
diff --git a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
index a0f2e2356cf81f14783c00394e0f4177a81cedab..9974ec443d4342baa69985e4cf38fad160299ebd 100644
--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -124,6 +124,12 @@ public abstract class BaseContainerBlockEntity extends BlockEntity implements Co
     @Override
     public void setItem(final int slot, final ItemStack itemStack) {
         this.getItems().set(slot, itemStack);
+        // Leaves start - grindstone overstacking
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && org.leavesmc.leaves.util.ItemOverstackUtils.CurseEnchantedBook.isCursedEnchantedBook(itemStack)) {
+            this.setChanged();
+            return;
+        }
+        // Leaves end - grindstone overstacking
         itemStack.limitSize(this.getMaxStackLeaves(itemStack)); // Leaves - item over-stack util
         this.setChanged();
     }
diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 88bdd463096294487d42bebdd245daa6cffcede9..4fdfb42689b2fe6c3aa7d150f2fe910c1f18f09f 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -114,6 +114,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public void setItem(final int slot, final ItemStack itemStack) {
         this.unpackLootTable(null);
         this.getItems().set(slot, itemStack);
+        // Leaves start - grindstone overstacking
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.allowGrindstoneOverstacking && org.leavesmc.leaves.util.ItemOverstackUtils.CurseEnchantedBook.isCursedEnchantedBook(itemStack)) {
+            return;
+        }
+        // Leaves end - grindstone overstacking
         itemStack.limitSize(this.getMaxStackLeaves(itemStack)); // Leaves - item over-stack util
     }
 
