From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Tue, 1 Jul 2025 01:29:07 +0800
Subject: [PATCH] Old Throwable Projectile tick order


diff --git a/net/minecraft/world/entity/projectile/ThrowableProjectile.java b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
index 799165c7e09b151c929fa386c33c6985aedb54d3..e96b4a0de255dcb968e89a1da1c29464c0f99836 100644
--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -45,6 +45,26 @@ public abstract class ThrowableProjectile extends Projectile {
     @Override
     public void tick() {
         this.handleFirstTickBubbleColumn();
+        // Leaves start - old-throwable-projectile-tick-order // TODO - ?
+        if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.oldThrowableProjectileTickOrder) {
+            super.tick();
+            HitResult hitResultOnMoveVector = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
+            if (hitResultOnMoveVector.getType() != HitResult.Type.MISS) {
+                this.preHitTargetOrDeflectSelf(hitResultOnMoveVector); // CraftBukkit - projectile hit event
+            }
+
+            Vec3 vec3 = this.getDeltaMovement();
+            double d = this.getX() + vec3.x;
+            double e = this.getY() + vec3.y;
+            double f = this.getZ() + vec3.z;
+            this.updateRotation();
+            this.applyInertia();
+            this.applyGravity();
+            this.setPos(d, e, f);
+            this.applyEffectsFromBlocks();
+            return;
+        }
+        // Leaves end - old-throwable-projectile-tick-order
         this.applyGravity();
         this.applyInertia();
         HitResult result = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
