From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sat, 23 May 2026 14:32:26 +0800
Subject: [PATCH] Fix piercing weapon can attack in spectator mode


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index fbdc03dc4a27ec23054705a4095485e0ce5c41c3..fbdd5d712a057269aea4972dd4b68e9b5d6a635e 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1978,7 +1978,7 @@ public class ServerGamePacketListenerImpl
                         }
 
                         PiercingWeapon piercingWeapon = itemInHand.get(DataComponents.PIERCING_WEAPON);
-                        if (piercingWeapon != null) {
+                        if (piercingWeapon != null && this.player.gameMode() != net.minecraft.world.level.GameType.SPECTATOR) { // Leaves - fix piercing weapon can attack in spectator mode
                             piercingWeapon.attack(this.player, EquipmentSlot.MAINHAND);
                         }
 
