From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 21:37:42 +0800
Subject: [PATCH] leaves paperServer File Patches


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 602d3ce2cd16405f35453ce12a207b70312a4a8a..39795db328b57f411df5a6c9e9666f2511976bae 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -218,7 +218,7 @@ public class PaperVersionFetcher implements VersionFetcher {
     }
 
     // Contributed by Techcable <Techcable@outlook.com> in GH-65
-    private static int fetchDistanceFromGitHub(final String branch, final String hash) {
+    protected static int fetchDistanceFromGitHub(final String branch, final String hash) { // Leaves - private -> protected
         try {
             final HttpURLConnection connection = (HttpURLConnection) URI.create("https://api.github.com/repos/%s/compare/%s...%s".formatted(PaperVersionFetcher.REPOSITORY, branch, hash)).toURL().openConnection();
             connection.setConnectTimeout(5000);
@@ -244,7 +244,7 @@ public class PaperVersionFetcher implements VersionFetcher {
         }
     }
 
-    private @Nullable Component getHistory() {
+    protected @Nullable Component getHistory() { // Leaves - private -> protected
         final VersionHistoryManager.@Nullable VersionData data = VersionHistoryManager.INSTANCE.getVersionData();
         if (data == null) {
             return null;
