From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:23 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6d0eb7c55856aff671204ddf390b2a1123b1441a..36f47ba41357773ef0b40b4a90040eec1756f7a8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -418,6 +418,8 @@ public final class CraftServer implements Server {
 
         CraftRegistry.setMinecraftRegistry(console.registryAccess());
 
+        org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.init(); // Leaves - Bytebuf API
+
         if (!Main.useConsole) {
             this.getLogger().info("Console input is disabled due to --noconsole command argument");
         }
@@ -2970,4 +2972,11 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Leaves end - replay mod api
+
+    // Leaves start - Bytebuf API
+    @Override
+    public org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator() {
+        return org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.allocator();
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index de83cbc631536aba76733e7cac0ea6b6780e3c14..e8da6f28df74145a99c20b6313b0cfad4fd7b512 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3402,4 +3402,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     public PlayerPostEffects postEffects() {
         return new PaperPlayerPostEffects(this.getHandle());
     }
+
+    // Leaves start - Bytebuf API
+    @Override
+    public void send(org.leavesmc.leaves.bytebuf.PacketType packetType, org.leavesmc.leaves.bytebuf.Bytebuf bytebuf) {
+        org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.sendPacket(this, packetType, bytebuf);
+    }
+
+    @Override
+    public @org.jetbrains.annotations.NotNull Object getChannel() {
+        return getHandle().connection.connection.channel;
+    }
+    // Leaves end - Bytebuf API
 }
