From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:23 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index eda24e42686f607f20fc33cb6ec7b4af196cc5d9..1a55163f9680f31cda6550d7199a8a81ed2b5398 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -416,6 +416,8 @@ public final class CraftServer implements Server {
 
         CraftRegistry.setMinecraftRegistry(console.registryAccess());
 
+        org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.init(); // Leaves - Bytebuf API
+
         if (!Main.useConsole) {
             this.getLogger().info("Console input is disabled due to --noconsole command argument");
         }
@@ -2982,4 +2984,11 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Leaves end - replay mod api
+
+    // Leaves start - Bytebuf API
+    @Override
+    public org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator() {
+        return org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.allocator();
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 2847fe96181dbede9e81c9c6a3314c355cf13013..35a08b7141600d3705e41b0248069dbbcaf020d3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3368,4 +3368,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         super.knockback(strength, directionX, directionZ);
         this.entity.hurtMarked = true;
     }
+
+    // Leaves start - Bytebuf API
+    @Override
+    public void send(org.leavesmc.leaves.bytebuf.PacketType packetType, org.leavesmc.leaves.bytebuf.Bytebuf bytebuf) {
+        org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.sendPacket(this, packetType, bytebuf);
+    }
+
+    @Override
+    public @org.jetbrains.annotations.NotNull Object getChannel() {
+        return getHandle().connection.connection.channel;
+    }
+    // Leaves end - Bytebuf API
 }
