From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:58:42 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b2f79fb7a0940242d2cafe8f604a449967ecfc68..b2377c61cdd0661409d5cfe91ebbd5a15cd70bf2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2990,4 +2990,11 @@ public final class CraftServer implements Server {
         return org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.allocator();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    @Override
+    public org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return org.leavesmc.leaves.config.InternalConfigProvider.INSTANCE;
+    }
+    // Leaves end - Config API
 }
